core/src/main/java/org/apache/calcite/plan/volcano/RelSet.java

```diff
@@ -72,6 +74,23 @@
   RelSet equivalentSet;
   RelNode rel;
 
+  /**
+   * The position indicator of rel node that is to be processed.
+   */
+  private int relCursor = 0;
+
+  /**
+   * The relnodes after applying logical rules and physical rules,
+   * before trait propagation and enforcement.
+   */
+  final Set<RelNode> seeds = new HashSet<>();
+
+  /**
```

Comment:
Does the order matter since it's a HashSet?

Reply:
No, it doesn't matter. It just stores the initial operators before trait enforcement.

Reply:
Maybe an array list would serve the purpose?

Reply:
That also works. But I need a fast way to check whether the node is a seed or not. See `planner.isSeedNode()`